File: [6 kyu] Two Sum/main.py
def two_sum(numbers, target):
    for i in range(len(numbers) - 1):
        for j in range(i + 1, len(numbers)):
            if numbers[i] + numbers[j] == target:
                return (i, j)

    return (0, 0)


def two_sum(numbers, target):
    nums = []

    for i, num in enumerate(numbers):
        nums.append((num, i))

    nums.sort()

    left, right = 0, len(nums) - 1

    while left < right:
        sum = nums[left][0] + nums[right][0]

        if sum == target:
            return (nums[left][1], nums[right][1])
        elif sum < target:
            left += 1
        else:
            right -= 1

    return (0, 0)


from collections import defaultdict


def two_sum(numbers, target):
    nums = defaultdict(int)

    for i, num in enumerate(numbers):
        nums[num] = i

    for i, num in enumerate(numbers):
        if target - num in nums and nums[target - num] != i:
            return (nums[target - num], i)

    return (0, 0)


from collections import defaultdict


def two_sum(numbers, target):
    nums = defaultdict(int)

    for i, num in enumerate(numbers):
        if target - num in nums:
            return (nums[target - num], i)

        nums[num] = i

    return (0, 0)

File: [6 kyu] Two Sum/main.rs
fn two_sum(numbers: &[i32], target: i32) -> (usize, usize) {
    for i in 0..numbers.len() - 1 {
        for j in i + 1..numbers.len() {
            if numbers[i] + numbers[j] == target {
                return (i, j);
            }
        }
    }

    (0, 0)
}

fn two_sum(numbers: &[i32], target: i32) -> (usize, usize) {
    let mut nums = Vec::new();

    for (i, num) in numbers.iter().enumerate() {
        nums.push((num, i));
    }

    nums.sort();

    let mut left = 0;
    let mut right = nums.len() - 1;

    while left < right {
        let sum = nums[left].0 + nums[right].0;

        if sum == target {
            return (nums[left].1, nums[right].1);
        } else if sum < target {
            left += 1;
        } else {
            right -= 1;
        }
    }

    (0, 0)
}

use std::collections::HashMap;

fn two_sum(numbers: &[i32], target: i32) -> (usize, usize) {
    let mut maps = HashMap::new();

    for (i, num) in numbers.iter().enumerate() {
        maps.entry(num).or_insert(i);
    }

    for (i, num) in numbers.iter().enumerate() {
        if let Some(&index) = maps.get(&(target - num)) {
            if i != index {
                return (index, i);
            }
        }
    }

    (0, 0)
}

use std::collections::HashMap;

fn two_sum(numbers: &[i32], target: i32) -> (usize, usize) {
    let mut maps = HashMap::new();

    for (i, num) in numbers.iter().enumerate() {
        if let Some(&index) = maps.get(&(target - num)) {
            return (index, i);
        }

        maps.entry(num).or_insert(i);
    }

    (0, 0)
}
